### PR TITLE
Centrifuge app: Support for UI after runtime changes

### PR DESCRIPTION
**Pull request by sophialittlejohn**
### Description

This pull request...
- #911

Outstanding:
- ~show financing form logic needs to be fixed~
- ~closedLoans need to be added to queries~
- ~ongoing loan and loan list (assets tab) doesn't display data propertly~
- ~asset detail page: priced assets throws error~
- ~probably more...~

### Approvals

- [ ] Dev

### CODE REVIEW

**Comment by EstherCentrifuge** [comment]
@sophialittlejohn  if it's deployed can we close the ticket?

**Comment by sophialittlejohn** [comment]
@EstherCentrifuge please read the description above, it clearly states which tasks are still outstanding before this PR can be merged. I will make sure you know when this is finished and ready for QA 😄

File: centrifuge-app/src/components/DataTable.tsx [hunk 1]
```diff
@@ -49,6 +49,11 @@ export const DataTable = <T extends Record<string, any>>({
   lastGroupIndex,
   defaultSortOrder = 'desc',
 }: Props<T>) => {
+  const colWithUniqueKey = React.useMemo(
+    () => columns?.map((col) => ({ ...col, id: Math.random().toString(36).substr(2, 9) })),
```
Comment (sophialittlejohn) [inline centrifuge-app/src/components/DataTable.tsx:53]: a bit of a hacky way to generate a unique id for jsx map to use as a key.. happy for any improvement suggestions

File: centrifuge-app/src/pages/Loan/FinanceForm.tsx [hunk 2]
```diff
@@ -134,50 +140,52 @@ export const FinanceForm: React.VFC<{ loan: LoanType }> = ({ loan }) => {
...
-                    label="Amount"
-                    errorMessage={meta.touched ? meta.error : undefined}
-                    secondaryLabel={`${formatBalance(roundDown(maxBorrow), pool?.currency, 2)} available`}
-                    disabled={isFinanceLoading}
-                    currency={getCurrencySymbol(pool?.currency)}
-                    onChange={(value: number) => form.setFieldValue('amount', value)}
-                    onSetMax={() => form.setFieldValue('amount', maxBorrow)}
-                  />
+        {loan.status === 'Active' &&
+          allowedToBorrow[loan?.loanInfo ? loan.loanInfo.type : config.defaultLoanType] &&
```
Comment (sophialittlejohn) [inline centrifuge-app/src/pages/Loan/FinanceForm.tsx:144]: we don't allow users to select with loan type they want to create as of now. Regardless `loanInfo` is only set when the loan is priced - not when it's created. This is mainly just a fallback that's used in a few places to make up for the way loans are created on chain without having to build opinions into cent-js

File: centrifuge-app/src/components/DataTable.tsx [hunk 3]
```diff
@@ -72,9 +77,9 @@ export const DataTable = <T extends Record<string, any>>({
     <Stack as={rounded && !lastGroupIndex ? Card : Stack}>
       <Shelf>
         {showHeader &&
-          columns.map((col, i) => (
+          colWithUniqueKey.map((col, i) => (
             <HeaderCol
-              key={`${col.header}-${i}`}
+              key={col.id}
```
Comment (onnovisser) [inline centrifuge-app/src/components/DataTable.tsx:82]: Fine to just use `i` for this I think. The parent component may not memoize `columns`, so `colWithUniqueKey` might change every render.

File: centrifuge-js/src/modules/pools.ts [hunk 4]
```diff
@@ -1117,39 +1154,109 @@ export function getPoolsModule(inst: CentrifugeBase) {
...
+        (api) => combineLatest([api.query.loans.activeLoans(poolId)]),
+        (api, [activeLoanValues]) => ({ api, activeLoanValues })
+      ),
+      switchMap(({ api, activeLoanValues }) => {
+        const activeLoanData = activeLoanValues.toJSON() as ActiveLoanDetilsData[]
+        const interestAccrualKeys = activeLoanData.map((activeLoan) => hexToBN(activeLoan.interestRatePerSec))
+        const $interestAccrual = api.query.interestAccrual.rate.multi(interestAccrualKeys).pipe(take(1))
+        return combineLatest([
+          api.query.loans.loan.entries(poolId),
+          api.query.loans.activeLoans(poolId),
```
Comment (onnovisser) [inline centrifuge-js/src/modules/pools.ts:1176]: these are retrieved twice now. could do a `of(activeLoanValues)`

File: centrifuge-app/src/components/DataTable.tsx [hunk 5]
```diff
@@ -101,8 +106,8 @@ export const DataTable = <T extends Record<string, any>>({
             key={keyField ? row[keyField] : i}
             tabIndex={onRowClicked ? 0 : undefined}
           >
-            {columns.map((col) => (
-              <DataCol style={{ flex: col.flex }} align={col?.align} key={`${col.header}-${i}`}>
+            {colWithUniqueKey.map((col) => (
```
Comment (onnovisser) [inline centrifuge-app/src/components/DataTable.tsx:109]: same here. `i` is fine, so we won't need `colWithUniqueKey` with its random ids

**Review by onnovisser — APPROVED** [review]
💯 🚢 🚢

**Pull request merged**